### PR TITLE
JS: fix a few FPs in js/exception-xss

### PR DESCRIPTION
**Pull request by erik-krogh**
I found and fixed definite FPs in `js/exception-xss` by going through the results i LGTM.com. 

A few things happened in this commit. 

- The "is safe property access" check from `DomBasedXss.qll` was moved into a Sanitizer class. In this was the `exception-xss` query automatically benefits from this sanitizer. 
- Additionally the `.length` property is added as a property that does not leak taint. 
- The `getAnUnlikelyToThrowMethodName` is introduced and used. Its a hand curated list of methods that does not leak their arguments when throwing an exception. Its a horrible way to remove false positives, but I think its the best we can do for now. Some of them catch multiple FPs (I think `Object.assign` was the worst offender). 
- I define that nodes that are `instanceof Xss::Shared::Sink` cannot throw exceptions. This is to remove duplicates from `js/xss`. 

[Performance is fine, and no change in results for the default benchmarks](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1578644570194).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll [hunk 1]
```diff
@@ -12,15 +12,39 @@ module ExceptionXss {
...
+  /**
+   * Gets the name of a method that does not leak taint from its arguments if an exception is thrown by the method.
+   */
+  private string getAnUnlikelyToThrowMethodName() {
+    result = "getElementById" or
+    result = "indexOf" or
+    result = "stringify" or
+    result = "assign" or
+    // fs methods. (The callback argument to the async functions are vulnerable, but its unlikely that the callback is the user-controlled part).
+    result = "existsSync" or
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll]: For these fs methods, you may be able to make use of the externs like we do here: https://github.com/Semmle/ql/blob/c4d21633215700dae90f148ec1ef1182787b59b3/javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll#L400-L418

File: javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll [hunk 2]
```diff
@@ -12,15 +12,39 @@ module ExceptionXss {
...
+    result = "exists" or
+    result = "writeFileSync" or
+    result = "writeFile" or
+    result = "appendFile" or
+    result = "appendFileSync" or
+    result = "pick" or
+    // log.info etc.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll]: Have you seen: https://github.com/Semmle/ql/blob/c4d21633215700dae90f148ec1ef1182787b59b3/javascript/ql/src/semmle/javascript/frameworks/Logging.qll#L17-L32?

File: javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll [hunk 3]
```diff
@@ -79,15 +104,15 @@ module ExceptionXss {
     }
 
     /**
-     * Get the parameter in the callback that contains an error. 
+     * Get the parameter in the callback that contains an error.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll]:
```suggestion
     * Gets the parameter in the callback that contains an error.
```

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 4]
```diff
@@ -51,6 +51,24 @@ module Shared {
...
+        or
+        // `$(location.hash)` is a fairly common and safe idiom
+        // (because `location.hash` always starts with `#`),
+        // so we mark `hash` as safe for the purposes of this query
+        pacc.getPropertyName() = "hash"
+        or
+        pacc.getPropertyName() = "length"
+      )
+    }
+  }
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll]: Minor: this class belongs to the `DomBasedXss` module IMO.

File: javascript/ql/test/query-tests/Security/CWE-079/tst.js [hunk 5]
```diff
@@ -318,3 +318,10 @@ function basicExceptions() {
 function handlebarsSafeString() {
 	return new Handlebars.SafeString(location); // NOT OK!	
 }
+
+function test2() {
```
Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-079/tst.js]: Lets two more tests (at least): one for the js/xss deduplication, and one for the newly vetted method names.

File: javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll]: Now I have.

**Pull request merged**